### PR TITLE
[SPARK-19211][SQL] Explicitly prevent Insert into View or Create View As Insert

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Currently we don't explicitly forbid the following behaviors:
1. The statement CREATE VIEW AS INSERT INTO throws the following exception:
```
scala> spark.sql("CREATE VIEW testView AS INSERT INTO tab VALUES (1, \"a\")")
org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: org.apache.hadoop.hive.ql.metadata.HiveException: at least one column must be specified for the table;
 scala> spark.sql("CREATE VIEW testView(a, b) AS INSERT INTO tab VALUES (1, \"a\")")
org.apache.spark.sql.AnalysisException: The number of columns produced by the SELECT clause (num: `0`) does not match the number of column names specified by CREATE VIEW (num: `2`).;
```

2. The statement INSERT INTO view VALUES throws the following exception from checkAnalysis:
```
scala> spark.sql("INSERT INTO testView VALUES (1, \"a\")")
org.apache.spark.sql.AnalysisException: Inserting into an RDD-based table is not allowed.;;
'InsertIntoTable View (`default`.`testView`, [a#16,b#17]), false, false
+- LocalRelation [col1#14, col2#15]
```

After this PR, the behavior changes to:
```
scala> spark.sql("CREATE VIEW testView AS INSERT INTO tab VALUES (1, \"a\")")
org.apache.spark.sql.catalyst.parser.ParseException: Operation not allowed: CREATE VIEW ... AS INSERT INTO;

scala> spark.sql("CREATE VIEW testView(a, b) AS INSERT INTO tab VALUES (1, \"a\")")
org.apache.spark.sql.catalyst.parser.ParseException: Operation not allowed: CREATE VIEW ... AS INSERT INTO;

scala> spark.sql("INSERT INTO testView VALUES (1, \"a\")")
org.apache.spark.sql.AnalysisException: `default`.`testView` is a view, inserting into a view is not allowed;
```

## How was this patch tested?

Add a new test case in `SparkSqlParserSuite`;
Update the corresponding test case in `SQLViewSuite`.

### CODE REVIEW

**Comment by jiangxb1987** [comment]
cc @gatorsmile @cloud-fan Please have a look at this when you have time, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -604,7 +604,14 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+        val newTable = EliminateSubqueryAliases(lookupTableFromCatalog(u))
+        // Inserting into a view is not allowed, we should throw an AnalysisException.
+        newTable match {
+          case v: View =>
+            u.failAnalysis(s"${v.desc.identifier} is a view, inserting into a view is not allowed")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Can we move this to `PreprocessTableInsertion`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 2]
```diff
@@ -128,6 +129,15 @@ case class CreateViewCommand(
     qe.assertAnalyzed()
     val analyzedPlan = qe.analyzed
 
+    // CREATE VIEW AS INSERT INTO ... is not allowed, we should throw an AnalysisException.
+    analyzedPlan match {
+      case i: InsertIntoHadoopFsRelationCommand =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: `_: InsertIntoHadoopFsRelationCommand`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 3]
```diff
@@ -128,6 +129,15 @@ case class CreateViewCommand(
...
+        throw new AnalysisException("Creating a view as insert into a table is not allowed")
+      case i: InsertIntoDataSourceCommand =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: The same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 4]
```diff
@@ -128,6 +129,15 @@ case class CreateViewCommand(
     qe.assertAnalyzed()
     val analyzedPlan = qe.analyzed
 
+    // CREATE VIEW AS INSERT INTO ... is not allowed, we should throw an AnalysisException.
+    analyzedPlan match {
+      case i: InsertIntoHadoopFsRelationCommand =>
+        throw new AnalysisException("Creating a view as insert into a table is not allowed")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: It will be nice to put a view name in the error message.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: shall we forbid all commands? e.g. `CREATE VIEW xxx AS CREATE TABLE ...` should also be disallowed right?

Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: The sql parser only allows `CREATE VIEW AS query` here, a query can only be a `SELECT ...` or `INSERT INTO ...` or a CTE, so perhaps we don't have to consider other commands here.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: hmmmm, why `INSERT INTO ...` is a query?

Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]:
```
queryNoWith
    : insertInto? queryTerm queryOrganization                                              #singleInsertQuery
    | fromClause multiInsertQueryBody+                                                     #multiInsertQuery
    ;
```
Seems we have mixed them together.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: can we fix it at parser side? cc @hvanhovell

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Hmmm... fixing this in grammar itself would a little bit of work (we also support multi-insert and that makes this harder). I suppose we could try to add a check in the `AstBuilder`.

Why not check this in the `CreateViewCommand`? That seems safer to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -604,7 +604,14 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+        val newTable = EliminateSubqueryAliases(lookupTableFromCatalog(u))
+        // Inserting into a view is not allowed, we should throw an AnalysisException.
+        newTable match {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: NIT In this case if would just write the following:
```scala
if (newTable.isInstanceOf[View]) {
  u.failAnalysis(s"${v.desc.identifier} is a view, inserting into a view is not allowed")
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 6]
```diff
@@ -128,6 +129,15 @@ case class CreateViewCommand(
     qe.assertAnalyzed()
     val analyzedPlan = qe.analyzed
 
+    // CREATE VIEW AS INSERT INTO ... is not allowed, we should throw an AnalysisException.
+    analyzedPlan match {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: This pattern match does not work with multi-inserts (hive feature). Those are represented using a `Union` of inserts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -604,7 +604,13 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+        val newTable = EliminateSubqueryAliases(lookupTableFromCatalog(u))
+        // Inserting into a view is not allowed, we should throw an AnalysisException.
+        if (newTable.isInstanceOf[View]) {
```
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: The rule `ResolveRelations` executes before `PreprocessTableInsertion`, so we can fail early here.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We prefer to doing all the error handling in the same rule. It can help us find the hole and maintain the codes. cc @cloud-fan

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: In fact I'm neutral with which rule this code block should be placed, but I feel this is not a typical "Preprocess" since the query failed and the process ends.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Just move it to preprocess insert.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 8]
```diff
@@ -484,6 +485,23 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("create view as insert into table") {
+    withView("testView") {
+      // Single insert query
+      val e1 = intercept[ParseException] {
+        sql(s"CREATE VIEW testView AS INSERT INTO jt VALUES(1, 1)")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Nit: remove string interpolator.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 9]
```diff
@@ -484,6 +485,23 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
...
+      }.getMessage
+      assert(e1.contains("Operation not allowed: CREATE VIEW ... AS INSERT INTO"))
+
+      // Multi insert query
+      val e2 = intercept[ParseException] {
+        sql(s"CREATE VIEW testView AS FROM jt INSERT INTO tbl1 SELECT * WHERE jt.id < 5 " +
+          s"INSERT INTO tbl2 SELECT * WHERE jt.id > 4")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Nit: remove the above two string interpolators

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 10]
```diff
@@ -484,6 +485,23 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("create view as insert into table") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Move it to `SparkSqlParserSuite`? 

If we do it here, it will be tested twice.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 11]
```diff
@@ -1329,6 +1329,16 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
     if (ctx.identifierList != null) {
       operationNotAllowed("CREATE VIEW ... PARTITIONED ON", ctx)
     } else {
+      // CREATE VIEW ... AS INSERT INTO is not allowed.
+      val query = ctx.query.queryNoWith
+      query match {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Nit: `ctx.query.queryNoWith match {`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 12]
```diff
@@ -1329,6 +1329,16 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+        case s: SingleInsertQueryContext if s.insertInto != null =>
+          operationNotAllowed("CREATE VIEW ... AS INSERT INTO", ctx)
+        case m: MultiInsertQueryContext =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Nit: `case MultiInsertQueryContext =>`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 13]
```diff
@@ -1329,6 +1329,15 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
     if (ctx.identifierList != null) {
       operationNotAllowed("CREATE VIEW ... PARTITIONED ON", ctx)
     } else {
+      // CREATE VIEW ... AS INSERT INTO is not allowed.
+      ctx.query.queryNoWith match {
+        case s: SingleInsertQueryContext if s.insertInto != null =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: when `s.insertInto` will be null?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 14]
```diff
@@ -604,7 +604,7 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+        i.copy(table = resolveRelation(EliminateSubqueryAliases(lookupTableFromCatalog(u))))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: why this change?

**Comment by cloud-fan** [comment]
LGTM except 2 questions

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 13 above)
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: For example, `CREATE VIEW v AS SELECT * FROM jt`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 14 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: When we try to insert into a view, the logical plan that `lookupTableFromCatalog()` returns is not expanded, so we have to perform `resolveRelation()` over the node.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 13 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: hmm, a select query is `SingleInsertQueryContext`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 14 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: so we will resolve something to `View` by doing this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 13 above)
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Yeah... You can see that in:
```
queryNoWith
    : insertInto? queryTerm queryOrganization                                              #singleInsertQuery
    | fromClause multiInsertQueryBody+                                                     #multiInsertQuery
    ;
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 14 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We will resolve the child of the view by doing this I think.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: why do we need to resolve the child of view? Once we see the pattern `Insert(View, ...)` we will throw exception, we don't care about whether the child of view is resolved or not.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Discussed with @cloud-fan and the concern is that in case the child of the view node is invalid(e.g. exists cyclic view reference or exceed max reference depth), we should still indicate that INSERT INTO VIEW is not allowed(instead of other error messages), so we should not resolve the child of the view, instead we throw an Exception here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 15]
```diff
@@ -604,7 +604,13 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+        val newTable = EliminateSubqueryAliases(lookupTableFromCatalog(u))
+        // Inserting into a view is not allowed, we should throw an AnalysisException.
+        if (newTable.isInstanceOf[View]) {
+          u.failAnalysis(s"${newTable.asInstanceOf[View].desc.identifier} is a view, inserting " +
+            s"into a view is not allowed")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Nit: `s"into a view is not allowed"` -> `"into a view is not allowed"`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 16]
```diff
@@ -604,7 +604,13 @@ class Analyzer(
...
+        }
+        i.copy(table = newTable)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: How about?

```Scala
        lookupTableFromCatalog(u).canonicalized match {
          case v: View =>
            u.failAnalysis(s"Inserting into a view is not allowed. View: ${v.desc.identifier}.")
          case other => i.copy(table = other)
        }
```

**Comment by cloud-fan** [comment]
LGTM, pending tests

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM too.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
